File: backend/app/schemas/workspace.py
from pydantic import BaseModel
from typing import Optional, List


class WorkspaceBase(BaseModel):
    name: str


class WorkspaceCreate(WorkspaceBase):
    pass


class WorkspaceUpdate(BaseModel):
    name: Optional[str] = None


class WorkspaceOut(WorkspaceBase):
    id: int
    owner_id: int
    created_at: Optional[str] = None

    class Config:
        from_attributes = True


class WorkspaceMemberOut(BaseModel):
    id: int
    workspace_id: int
    user_id: int
    role: str
    user_name: Optional[str] = None
    user_avatar: Optional[str] = None

    class Config:
        from_attributes = True


class WorkspaceDetailOut(WorkspaceOut):
    members: List[WorkspaceMemberOut] = []
    member_count: int = 0

    class Config:
        from_attributes = True


class WorkspaceSwitchRequest(BaseModel):
    workspace_id: int

File: frontend/api/workspace.ts
import { api } from "./api";

export interface WorkspaceItem {
  id: number;
  name: string;
  owner_id: number;
  created_at?: string;
  member_count: number;
  role: string;
}

export interface WorkspaceMember {
  id: number;
  workspace_id: number;
  user_id: number;
  role: string;
  user_name: string;
  user_avatar: string;
}

export interface WorkspaceDetail {
  id: number;
  name: string;
  owner_id: number;
  created_at?: string;
  members: WorkspaceMember[];
  member_count: number;
}

export const getUserWorkspaces = async (
  userId: number
): Promise<WorkspaceItem[]> => {
  const res = await api.get("/workspace", {
    params: { user_id: userId },
  });
  return res.data;
};

export const getWorkspaceDetail = async (
  workspaceId: number
): Promise<WorkspaceDetail> => {
  const res = await api.get(`/workspace/${workspaceId}`);
  return res.data;
};

export const createWorkspace = async (
  userId: number,
  name: string
): Promise<WorkspaceItem> => {
  const res = await api.post(
    "/workspace",
    { name },
    { params: { user_id: userId } }
  );
  return res.data;
};

export const updateWorkspace = async (
  workspaceId: number,
  userId: number,
  name: string
): Promise<{ id: number; name: string; owner_id: number }> => {
  const res = await api.put(
    `/workspace/${workspaceId}`,
    { name },
    { params: { user_id: userId } }
  );
  return res.data;
};

export const deleteWorkspace = async (
  workspaceId: number,
  userId: number
): Promise<{ message: string; active_workspace_id: number | null }> => {
  const res = await api.delete(`/workspace/${workspaceId}`, {
    params: { user_id: userId },
  });
  return res.data;
};

export const switchWorkspace = async (
  userId: number,
  workspaceId: number
): Promise<{
  message: string;
  workspace: { id: number; name: string; owner_id: number };
}> => {
  const res = await api.post(
    "/workspace/switch",
    { workspace_id: workspaceId },
    { params: { user_id: userId } }
  );
  return res.data;
};